reactioncommerce/reaction
# Reaction Endpoints

Adds new `Endpoints` module to Reaction on the server.

I'm using it to handle webhook requests that will be incoming from Shopify, but it can be used to create any kind of HTTP endpoint that you might need.

It's based on Stubailo's [JSON Routes](https://github.com/stubailo/meteor-rest/tree/master/packages/json-routes) meteor package, but with the the meteor package stuff removed, and updated to ES6 and with a few node module updates.

It uses Meteor's [WebApp](https://docs.meteor.com/packages/webapp.html) module internally and Sencha Lab's [Connect](https://github.com/senchalabs/connect) middleware framework.

It permits you to write code like the following examples in reaction (even in a plugin).

Example `POST` endpoint
```
import { Reaction } from "/server/api";

Reaction.Endpoints.add("post", "/shopify-webhooks/update-product", function (req, res) {
  console.log("req", req); // The request object
  console.log("headers", req.headers); // HTTP headers sent along with the request
  console.log("body", req.body); // request body

  Reaction.Endpoints.sendResponse(res, { // send a response
    code: 200, // optional, default is 200 ok
    headers: {}, // optional, dictionary of response headers
    data: { // optional, data to respond with. 
      id: "123"
    }
  });
});
```

Example `GET` endpoint
```
Reaction.Endpoints.add("get", "/test-endpoint", function (req, res) {
  console.log("req", req);
  console.log("headers", req.headers);
  console.log("body", req.body);

  Reaction.Endpoints.sendResponse(res, {
    // code: 200, default code 200, no need to add this unless you want to send
    // a different code
    data: {
      message: "hello world",
      foo: "bar"
    }
  });
});
```

I use [ngrok](https://ngrok.com/) to test with webhooks on localhost.

**To Test**
1. Add a file to the server and make sure it's being imported
2. Write an endpoint, such as the example `GET` endpoint above.
3. Hit it with a `curl` request using `ngrok`
e.g.
`curl http://yourhash.ngrok.io/test-endpoint`
4. See the message
